### PR TITLE
Ensure we do not accidentally override validation rules.

### PR DESCRIPTION
**Pull request by nicosp**
Avoid replacing validation rules by accident. Similar approach to the recent PR to protect rules from accidental replacement using the same name.

### CODE REVIEW

**Comment by ADmad** [comment]
Please rebase your branch onto upstream 5.next again to get rid of the failure on php 8.4

File: src/Validation/ValidationSet.php [hunk 1]
```diff
@@ -138,12 +139,16 @@ public function rules(): array
...
      * @return $this
+     * @throws \Cake\Core\Exception\CakeException If a rule with the same name already exists
      */
     public function add(string $name, ValidationRule|array $rule)
     {
         if (!($rule instanceof ValidationRule)) {
             $rule = new ValidationRule($rule);
         }
+        if (array_key_exists($name, $this->_rules)) {
+            throw new CakeException('A validation rule with the same name already exists');
```
Comment (ADmad) [inline src/Validation/ValidationSet.php]:
```suggestion
            throw new CakeException("A validation rule with the name `{$name}` already exists");
```

**Comment by nicosp** [comment]
> Please rebase your branch onto upstream 5.next again to get rid of the failure on php 8.4

I merged the current 5.next. Am I missing something?

**Comment by ADmad** [comment]
> I merged the current 5.next

That created an unnecessary extra commit :)

File: tests/TestCase/Validation/ValidationRuleTest.php [hunk 2]
```diff
@@ -178,4 +180,14 @@ public function testGet(): void
...
     }
+
+    public function testAddDuplicateName(): void
+    {
+        $rules = new ValidationSet();
+        $rules->add('myUniqueName', ['rule' => fn () => false]);
+
+        $this->expectException(CakeException::class);
+        $rules->add('myUniqueName', ['rule' => fn () => true]);
+        $this->fail('Exception not thrown');
```
Comment (ADmad) [inline tests/TestCase/Validation/ValidationRuleTest.php]: Is this fail() call necessary? I am pretty sure the test will fail if the exceptException() isn't met.

Comment (nicosp) [reply tests/TestCase/Validation/ValidationRuleTest.php]: It's not. Removing it

**Comment by ADmad** [comment]
With this change we also need to add a `ValidationRule::has($name)` method, it would be identical to `ValidationRule::offsetExists()`.

**Review by ADmad — APPROVED** [review]

**Pull request merged**